File: editor/en/[310]Minimum Height Trees.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var findMinHeightTrees = function (n, edges) {
    if (n === 1) {
        return [0];
    }

    const graph = new Array(n).fill(0).map(() => []);
    const degree = new Array(n).fill(0);

    for (const [u, v] of edges) {
        graph[u].push(v);
        graph[v].push(u);
        degree[u]++;
        degree[v]++;
    }

    const leaves = [];

    for (let i = 0; i < n; i++) {
        if (degree[i] === 1) {
            leaves.push(i);
        }
    }

    let remainingNode = n;

    while (remainingNode > 2) {
        const size = leaves.length;
        remainingNode -= size;

        for (let i = 0; i < size; i++) {
            const leaf = leaves.shift();

            for (const neighbor of graph[leaf]) {
                degree[neighbor]--;

                if (degree[neighbor] === 1) {
                    leaves.push(neighbor);
                }
            }
        }
    }

    return leaves;
};
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[310]Minimum Height Trees.py
# leetcode submit region begin(Prohibit modification and deletion)
from collections import defaultdict, deque


class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]

        graph = defaultdict(list)
        degree = [0] * n

        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
            degree[u] += 1
            degree[v] += 1

        leaves = deque(i for i in range(n) if degree[i] == 1)
        remaining_node = n

        while remaining_node > 2:
            size = len(leaves)
            remaining_node -= size

            for _ in range(size):
                leaf = leaves.popleft()

                for neighbor in graph[leaf]:
                    degree[neighbor] -= 1

                    if degree[neighbor] == 1:
                        leaves.append(neighbor)

        return list(leaves)
# leetcode submit region end(Prohibit modification and deletion)
